### PR TITLE
parser: complete Restore for CreateUserStmt

### PR DESCRIPTION
**Pull request by spongedu**
### What problem does this PR solve? 
in #269 I've refined the `CREATE USER` syntax in parser. This pr completes the `Restore` part for it.

### What is changed and how it works?
Define related Asts in `ast/misc.go` and do the related construction work in parser.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

### CODE REVIEW

File: ast/misc.go [hunk 1]
```diff
@@ -901,14 +901,114 @@ func (n *UserSpec) EncodedPassword() (string, bool) {
...
+	case Cipher:
+		ctx.WriteKeyWord("CIPHER ")
+		ctx.WriteString(t.Value)
+	case Issuer:
+		ctx.WriteKeyWord("ISSUER ")
+		ctx.WriteString(t.Value)
+	case Subject:
+		ctx.WriteKeyWord("CIPHER")
+		ctx.WriteString(t.Value)
+	}
```
Comment (kennytm) [inline ast/misc.go]: Add a default case and return error.

```suggestion
	default:
		return errors.Errorf("Unsupported TslOption.Type %d", t.Type)
	}
```

File: ast/misc.go [hunk 2]
```diff
@@ -901,14 +901,114 @@ func (n *UserSpec) EncodedPassword() (string, bool) {
...
+	switch r.Type {
+	case MaxQueriesPerHour:
+		ctx.WritePlain("MAX_QUERIES_PER_HOUR ")
+	case MaxUpdatesPerHour:
+		ctx.WritePlain("MAX_UPDATES_PER_HOUR ")
+	case MaxConnectionsPerHour:
+		ctx.WritePlain("MAX_CONNECTIONS_PER_HOUR ")
+	case MaxUserConnections:
+		ctx.WritePlain("MAX_USER_CONNECTIONS ")
+	}
```
Comment (kennytm) [inline ast/misc.go]: (Ditto, add a default)

File: ast/misc.go [hunk 3]
```diff
@@ -901,14 +901,114 @@ func (n *UserSpec) EncodedPassword() (string, bool) {
...
+
+type ResourceOption struct {
+	Type  int
+	Count int64
+}
+
+func (r *ResourceOption) Restore(ctx *RestoreCtx) error {
+	switch r.Type {
+	case MaxQueriesPerHour:
+		ctx.WritePlain("MAX_QUERIES_PER_HOUR ")
```
Comment (kennytm) [inline ast/misc.go]: These should be keywords?

```suggestion
		ctx.WriteKeyWord("MAX_QUERIES_PER_HOUR ")
```

File: ast/misc.go [hunk 4]
```diff
@@ -901,14 +901,114 @@ func (n *UserSpec) EncodedPassword() (string, bool) {
...
+
+type PasswordOrLockOption struct {
+	Type  int
+	Count int64
+}
+
+func (p *PasswordOrLockOption) Restore(ctx *RestoreCtx) error {
+	switch p.Type {
+	case PasswordExpire:
+		ctx.WritePlain("PASSWORD EXPIRE")
```
Comment (kennytm) [inline ast/misc.go]: (Ditto, `WriteKeyWord`)

File: ast/misc.go [hunk 5]
```diff
@@ -901,14 +901,114 @@ func (n *UserSpec) EncodedPassword() (string, bool) {
...
+		ctx.WritePlain("PASSWORD EXPIRE NEVER")
+	case PasswordExpireInterval:
+		ctx.WritePlainf("PASSWORD EXPIRE NEVER %d DAY", p.Count)
+	case Lock:
+		ctx.WritePlain("ACCOUNT ")
+		ctx.WriteKeyWord("LOCK")
+	case Unlock:
+		ctx.WritePlain("ACCOUNT ")
+		ctx.WriteKeyWord("UNLOCK")
+	}
```
Comment (kennytm) [inline ast/misc.go]: (Ditto, `default:`)

File: ast/misc.go [hunk 6]
```diff
@@ -929,6 +1029,42 @@ func (n *CreateUserStmt) Restore(ctx *RestoreCtx) error {
...
+		if i != tslOptionLen {
+			ctx.WriteKeyWord(" AND ")
+		}
+		if err := n.TslOptions[i-1].Restore(ctx); err != nil {
+			return errors.Annotatef(err, "An error occurred while restore CreateUserStmt.TslOptions[%d]", i)
+		}
+	}
+
+	if len(n.ResourceOptions) != 0 {
+		ctx.WriteKeyWord(" WITH ")
```
Comment (kennytm) [inline ast/misc.go]:
```suggestion
		ctx.WriteKeyWord(" WITH")
```

so that...

File: ast/misc.go [hunk 7]
```diff
@@ -929,6 +1029,42 @@ func (n *CreateUserStmt) Restore(ctx *RestoreCtx) error {
...
+	}
+
+	for i, v := range n.ResourceOptions {
+		if i != 0 {
```
Comment (kennytm) [inline ast/misc.go]: ... this doesn't need to be conditional 🙂 

```suggestion
```

File: ast/misc.go (hunk 2 above)
Comment (spongedu) [reply ast/misc.go]: ok

File: ast/misc.go (hunk 3 above)
Comment (spongedu) [reply ast/misc.go]: Should the `UnReservedKeyword`s be restored using `WriteKeyWord` ?

**Comment by spongedu** [comment]
@kennytm PTAL

File: ast/misc.go (hunk 3 above)
Comment (kennytm) [reply ast/misc.go]: Yes, every word that acts like a keyword (being unquoted and case-insensitive) should be restored using `WriteKeyWord`.

**Pull request merged**